resources/rest-service/cloudify/install_agent.py

```diff
@@ -68,14 +70,31 @@ def run(self, command, execution_env=None):
             self.logger.error(out)
             self.logger.error('Stderr:')
             self.logger.error(err)
-            raise Exception()
+            raise CommandExecutionException(command, err, out, p.returncode)
 
-    def download(self, url, destination=None):
+    def download(self, url, destination=None, verify_certificate=True,
+                 certificate_file=None):
         self.logger.debug('Retrieving file from {0}'.format(url))
-        if destination is None:
-            fh_num, destination = tempfile.mkstemp()
-            os.close(fh_num)
-        urllib.urlretrieve(url, destination)
+
+        verify = False
+        if verify_certificate:
+            if certificate_file:
+                verify = certificate_file
+            else:
+                verify = True
+        response = requests.get(url, stream=True, verify=verify)
+
+        if destination:
+            destination_file = open(destination)
+        else:
+            destination_file = tempfile.NamedTemporaryFile(delete=False)
+            destination = destination_file.name
+
+        with destination_file as f:
+            for chunk in response.bytes_stream(8192):
+                f.write(chunk)
+            f.flush()
```

Comment:
the `with` statement closes it. `flush` isn't technically needed at this point but rest-client from where i copied this does it